manticore/utils/mappings.py

```diff
@@ -46,10 +46,15 @@ def mmap(fd, offset, size):
 
     assert size > 0
 
-    result = mmap_function(0, size, prot, flags, fd, offset)
-    return ctypes.cast(result, ctypes.POINTER(ctypes.c_char))
+    aligned_offset = offset & ~0xfff
+    result = mmap_function(0, size, prot, flags, fd, aligned_offset)
```

Comment:
```suggestion
    result = mmap_function(0, size + offset - aligned_offset, prot, flags, fd, aligned_offset)
```

Assume that:
* offset = 0x1001
* size/length = 10

The `alligned_offset` is 0x1000. We mmap 10 bytes so the `result` will be an area of 10 bytes***.

Later, we return a pointer to `result+offset-aligned_offset` which effectively is `result+1`. Since `result` is of 10 bytes, we will end up pointing to an area of 9 bytes, while we wanted a 10 bytes area.

And that's the reason why we need to adjust `size` argument properly.

I have based my assumptions on the example taken from `man mmap` where they call `mmap` like this (`pa_offset` is the aligned size, `offset` is size):
```c
addr = mmap(NULL, length + offset - pa_offset, PROT_READ,
                       MAP_PRIVATE, fd, pa_offset);
```

and then they use the memory as (so it is similar to us returning the adjusted pointer):
```c
write(STDOUT_FILENO, addr + offset - pa_offset, length);
```
---
*** well, `mmap` will probably return a whole memory page (by default 4 kB on most or all kernels) but we probably can't assume that.

Reply:
> We mmap 10 bytes so the result will be an area of 10 bytes***.

I think your code already in place that aligns the `size` on the page size will have changed the `size` value to 0x1000, so you won't be mmap'ing 10 bytes but 0x1000.

Reply:
Why 0x1000? size + offset - aligned_offset = 10 + 0x1001 - 0x1000 = 10 + 1 = 11, so yeah we mmap more bytes but we return pointer to the result+1.

Reply:
Please consider the following example:
```
def mmap(fd, offset, size):
    if size & 0xfff != 0:
        size = (size & ~0xfff) + 0x1000
    assert size > 0

    aligned_offset = offset & ~0xfff
    print("mmap_function(0, %#x, prot, flags, fd, %#x)" % (size, aligned_offset))
    result = 0x12340000 # page aligned address returned by mmap()
    return result + offset - aligned_offset

def munmap(address, size):
    aligned_address = address & ~0xfff
    print("munmap_function(%#x, %#x)" % (aligned_address, size))

offset = 0x1
size = 0x10

address = mmap(0, offset, size)
print("result is %#x" % address)
munmap(address, size)
```

and the corresponding output:
```
mmap_function(0, 0x1000, prot, flags, fd, 0x0)
result is 0x12340001
munmap_function(0x12340000, 0x10)
```

We're trying to map 0x10 bytes from a file starting at offset 0x1. You can see that we end up calling mmap() with a offset of 0 and a size of 0x1000. The address returned by mmap() is the address at which the beginning of the file is mapped, so we need to add 0x1 to the pointer returned to the user.

When the user tries to free the memory using this pointer, we align it back to recover the value returned by mmap(). We can use the original size because munmap() will unmap the whole page.

Reply:
Oh, I am sorry, I missed the operations above that deal with `size`. So yeah, this is correct :).

:+1:

Reply:
I did more thinking and it turns out that you were right!

Consider an offset of 0xfff and a size of 0x2. You would need to mmap (and then unmap) 0x2000 bytes starting at offset 0x0, but the current code would only mmap (and unmap) 0x1000.

I've pushed two other commits that fix this issue (with comments!).